### PR TITLE
[Sema] TypeWrappers: If type wrapper comes from `.swiftinterface` use synthesized decls

### PR DESCRIPTION
**Pull request by xedin**
Swift interfaces have both attribute and all publicly accessible
synthesized members, which means that the synthesis needs to be
taught to lookup existing synthesized declarations for `$storage`,
`$Storage`, and `init(storageWrapper:)` if request comes for
declaration that belongs to a Swift interface file.

Note that this PR also allows type wrappers if they appear in the
swiftinterface file because not all use situations could be detected
by `usesFeature*` methods i.e. existential value uses of type wrapped
protocol.

Resolves:  rdar://103425758

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/CodeSynthesis.cpp [hunk 1]
```diff
@@ -1657,14 +1657,29 @@ ConstructorDecl *SynthesizeTypeWrappedTypeStorageWrapperInitializer::evaluate(
...
     return nullptr;
 
+  auto &ctx = wrappedType->getASTContext();
+
+  // .swiftinterfaces have both attribute and a synthesized member
+  // (if it's public), so in this case we need use existing declaration
+  // if available.
+  {
+    auto parentSF = wrappedType->getDeclContext()->getParentSourceFile();
+    if (parentSF && parentSF->Kind == SourceFileKind::Interface) {
```
Comment (xymus) [inline lib/Sema/CodeSynthesis.cpp:1667]: I'm wondering if it's worth restricting this logic to swiftinterfaces. Could it be useful for the user to define custom implementations for the type wrapper's backing logic? Keeping a similar behavior between sources and swiftinterface would be nice too.

File: test/Serialization/type_wrapper_in_swiftinterface.swift [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    get { fatalError() }
+  }
+
+  public subscript<V>(propertyKeyPath _: KeyPath<W, V>, storageKeyPath path: WritableKeyPath<S, V>) -> V {
+    get { fatalError() }
+    set { }
+  }
+}
+
+//@Wrapper
```
Comment (xymus) [inline test/Serialization/type_wrapper_in_swiftinterface.swift:49]: Forgotten comment?

**Review by xymus — APPROVED** [review]
This looks good!

File: test/Serialization/type_wrapper_in_swiftinterface.swift (hunk 2 above)
Comment (xedin) [reply test/Serialization/type_wrapper_in_swiftinterface.swift:49]: Oops, thank you!

File: lib/Sema/CodeSynthesis.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CodeSynthesis.cpp:1667]: This is something we intentionally don't want users to mess with.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

File: test/Serialization/type_wrapper_in_swiftinterface.swift [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+// RUN: %empty-directory(%t/sdk)
+// RUN: split-file %s %t/src
+
+/// Build the library.
+// RUN: %target-swift-frontend -emit-module %t/src/PublicModule.swift \
+// RUN:   -module-name PublicModule -swift-version 5 -enable-library-evolution \
+// RUN:   -emit-module-path %t/PublicModule.swiftmodule \
+// RUN:   -emit-module-interface-path %t/PublicModule.swiftinterface \
+// RUN:   -emit-private-module-interface-path %t/PublicModule.private.swiftinterface \
+// RUN:   -enable-experimental-feature TypeWrappers
```
Comment (drodriguez) [inline test/Serialization/type_wrapper_in_swiftinterface.swift:11]: Many of the tests that enable `TypeWrappers` seems to include `REQUIRES: asserts`. This test does not, and if one runs the test suite in a non-assert compiler, one gets the error "experimental feature 'TypeWrappers' cannot be enabled in a production compiler". Would it be correct adding `REQUIRES: asserts` also to this test?

**Pull request merged**